Metric

cryogenian opened the pull request:
fixes #951 

![with-label](https://cloud.githubusercontent.com/assets/10245930/18344149/d2e89920-75be-11e6-9ef8-deb94924efb2.png)
![without-label](https://cloud.githubusercontent.com/assets/10245930/18344148/d2d5917c-75be-11e6-8910-3aa3b69b97a2.png)

@jdegoes I can add vertical align if you wish. Here I thought about using draftboard gaps and empty regions to emulate middle alignment like 
![middle](https://cloud.githubusercontent.com/assets/10245930/18344243/1f16f058-75bf-11e6-93fe-4afeb24066d2.png)

@beckyconning Please review

jdegoes commented:
The metric should be centered horizontally and almost centered vertically (but slightly offset up to accommodate for label so the overall group looks perfectly vertically centered). See the examples I sent previously.

cryogenian commented:
Added vertical align 
![centered](https://cloud.githubusercontent.com/assets/10245930/18363250/fe87d692-7611-11e6-85e3-8f366f8a1b7e.png)

beckyconning commented:
The vertical align makes a big difference. Sorry for the delay in review. I'm about half way through and I'll finish it off early tomorrow.

jdegoes commented:
@cryogenian I think you should show a more realistic mock, as I think the spacing may still be off.

e.g.

```
92%
Conversion
```

And

```
2,302
Unique Visitors
```

jdegoes commented:
![](data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBxASEhUSEhIWFRUVGBYWGBgWGSAXFxgYGBUYFxgXFhUYKCggHRslHRcVIjEiJSorLi4uGB8zODMtNygtLisBCgoKDg0OGhAQGy0lHyU1LS8uKystLi4tLys3LS0tLy8tLS0tLS0tLS8vLS0wMC0tKy0tLy0tLS0tLS0tLS0tLf/AABEIAIkBcAMBIgACEQEDEQH/xAAcAAACAgMBAQAAAAAAAAAAAAAACAUGAQIEAwf/xABJEAABAgMDBAwNAwMEAQUAAAABAAIDBBEFIZIGEjFSExUWIjRBUVSRs8HRFzIzYXFyc4GCk6Gx0wcUdCNCsiRi4fDxQ0RTotL/xAAZAQEBAQEBAQAAAAAAAAAAAAAAAQIDBAX/xAArEQEBAAIBAwIEBQUAAAAAAAAAAQIREgMhMUFhEzJRgRQicbHwBJGh0eH/2gAMAwEAAhEDEQA/APlgaTQAVJuAF5J5AFLDJme5tFwrlsPhUv7eB1rEx9kWfCiQ85wJNSNJH2UtC87mZ7m0XCs7mZ7m0XCmS2mgapxFG08DVOIqcgtoyZnuaxcKNzM9zaLhTJbTwNU4ijaeBqnEU5BbRkzPc2i4Ubmp7m0XCmS2mgapxFG00DVOIpyC1uyZnuaxcKzuZnuaxcKZPaaBqnEUbTQNU4inILZuZnuaxcKy3Jme5tFwpktpoGqcRRtNA1TiKcgt25me5rFwoGTM9zaLhTI7TwNU4ijaeBqnEU5BbX5NzwBP7aLdf4tfoFEDzcaZq2JGHDa0sFCXU0k3UJ40ukpwpn8hvXBWUe4yZnqcGi4VkZMz3NouFMJY9nw4jC54qc4jSRdQcnpXdtPA1TiKmwtpyZnuaxcKzuZnubRcKZHaeBqnEUbTwNU4inILacmZ7msXCsDJme5tFwpk9poGqcRRtNA1TiKcgtu5me5rFwrAyZnuaxcKZPaaBqnEUbTQNU4inILaMmZ7m0XCs7mZ7m0XCmR2mgapxFG08DVOIpyC2HJme5rFwoOTM9zWLhTJ7TQNU4ijaaBqnEU5BbhkzPc2i4UOyanqcGi4UyO08DVOIrhtiQhw2BzBQ5wGkm6h5fQnILKVKDJueIBErFvv8Wn0K8Zg/wCqf/If1xTF2PIQ4jXF4qQ6mkjiHIrboL4MmZ7m0XCsDJme5rFwpk9p4GqcRRtNA1TiKnILbuZnubRcKBkzPc1i4UyW08DVOIo2ngapxFOQW3czPc2i4UDJme5tFwpktp4GqcRRtNA1TiKcgtu5qe5tFwrByZnq8Fi4Uye08DVOIo2mgapxFOQWzczPc1i4UbmZ7msXCmT2mgapxFG00DVOIpyC2tyZnubRcKzuZnuaxMKZHaaBqnEUbTwNU4inILduZnubRcKingioIoRcQbiDyEcqZq1rPhQ4ec0UNQNJP3S4W3wmY9vH656suxiw+FS3t4HWtTNZP+S+I9iWWxOFS3t4HXMTNWB5L4ndimQkULzmphkNjojzmsY0uceRrRUm6/QFT4OXTY09Bl5YB0J9Q97gQSc1x3gNKC4aQsquiFWMq8pIsGLClZWG2JMRbxneK1tSASARyO47g1eNg5TTH7l0nPQ2Q4oaXtczxHADOOkniBNf9puCC2oVAGVtpTAix5SXhmXhE+PUveGipIoRfShoNFeNWzJu2WzkuyO0Ztahza1zXA0IrxjjHmIQSaEIQCEIQCEIQRWUXiN9bsKWuU4Uz27euCZTKLxG+v2FLXKcKZ/Ib1wWsUMtk75I+ufs1SajMnfJH1z9mrumnPDHFgq6lwPGVPNLdTb1Qq/M2pOQxnPhMaNHv9zl0TVqxGshjMBixB4vEPcunwcu3j+7h+Jw773NeyYQoeVtOMIohR2AF3ilv/fMVptjMvfEbCYxzWOLan/z5k+Dl7L+Ix16/TWu6bQoicno8OBnua1r84CmkU6V6WFaTozXZ1M5p4rrjo7VL0suPL0WdfG5zD1vdJoVdiW7E2fMbm5meG6L6Vob1Y1M+ncNb9V6fWx6m+PowhCFh1CjMovJD1x9nKTUZlD5IeuP8XIFomOFP/kP64plcnfEd6/YEtcfhT/5D+uKZTJ3xHev2BaqJVC0mIzWNc92hoLj6AKlVTdBOuY6YZBZsLSdPjUGk1r9guWWcxS5SLchV6fyidsEGLBaC6M7NDXX0N4Iup/dS9e8hHtAxGiLChtZfnFpvFxpS88dE5zeocomkLymo7YbHPdoaC4+4Kr7oJ0wzMiCzYQdF+dStK1ry3Vp7kyzmPkuUi2oUBaeUVGQdgbnxI4BaDoA0X0463e4rFnWzHEcS80xrXOFWOZoOm46eQ9CfEx3o5TawIVXNuzcZ0QysJhhwzSrtLqcl46FMWFagmYQiAUIOa4chHJ5rwUx6kt1CZSpBCELbSOt/wAl8Q7Us1tcJmfbx+uemZt/yXxDtSzW0f8AUzPt4/XPWsUYsPhMv7eB1rUzWT/kviPYllsM/wCqlvbwOtamasDyXxO7EyEg9gcCCAQRQg3gg6QRyL59a8uyHbcmGMawbGDRoDRX+tfQe5X6YmGQ25z3BrbhVxoKm4C9V2dlJKJOQpt8xmxIW8a3PYGupW+mk+U4jyLKomfObb0Eu0OhUbgiD7g9K5crWGJa0NkO9wlogNOImFHpXE3pCsVvWfIT7WEx2hzDvIkJ7c4b5gpW8EZzofvc2mlc+TNmWfLOMVk0I0WJvdkfEa5xrQ5ractW9IVHH+m8wwWbEJIox0XP829Dr/cVv+kjHCScToMVxHoDWA/UFaWlkfZ5jRAJp0DO38SC2I0Np4xJadAoa33DiorNY7pSGxsCA9maxtQ0OBIbxud6TWpPHVBJIUbCt6UdWkdlAc2pcADvWuJaeMUe2/lNF3wozXVzXB1DQ0NaGgNLvMR0qDdCEIBCEIIrKLxG+t2FLVKcLZ/Ib1wTK5ReI31uwpapThbPbt60LWKGXye8kfXP+LVJqMye8kfXP+LVIxYgaC46ACT7lkQsU7NNhn9kIVpyuu7adC7LVjwYVIj2hz9DNb3HiF/1XPKmDCe6IHPOyAOvFwznGmjRfcs2jIwYxZEdELc5oDRUCvHcDx3/AGXp7cpveo8c5cMrNcrd/wCmknKRYkQTEbe5o3reQef6rwlLIhxGvcI5deTVu9ANK1IK2gSUBj2uEd5oa6QW3cpW0eyIF72xSxjjQgOGadN30Ny1y1fNn2Y4bnyy+d7v+fp/OyOdMvfKODjXNiNAJ5KVpVbbKZd4eNESECPWzf8A9D6qXjScvsWw7IGgEG5wrXirXlWLRlIDhDhvfQsFBQitM3+7kFAtTq42612u2b0M5N7m5rV9+6DbL5n7YnS92cfe5tPormoyPKQYroZD/E8UNIvpQ0P0UkuHWz56+/7vT/T9L4fL6dv2CEIXF6QozKLyQ9cfZyk1GZReSHrj7OQLTH4U/wDkP64plMnfEd6/YEtcxwp/8h/XFMpk74jvX7AtVElHY0tIeAWkHOropx18yqsWZfMgy0mwMgi50SlBQmpDR5+k+ZWG2AwwIgiOLWltHFumhuurcq0LAlGszv3EZtxdm1aHUFakNpfoN400uXn6m96jGW3RP2PBBlpf9xsZZUtFN84uIq4O0A1aVzxoBk5yC2FEe5sUgOa410uzan718xXRHs6UMuwF8RwY9zGPuz84uJIropUGleTzr3syzZWC/ZXRi+IADWK4BzagUq03g0I08qzx79vZNO3KVwdKxg0gkNqaGtwcCa+4FRsCK3asmt2xub8VSKdKLCgQmxph5cAyKc1oe5tX1JJIA/tJNBy1C5Jmx5Nj3sMWKGsczOhjxav8WnLdpOkJbb+b7F35cFltLY0kXaC00r53xKf5N6VLZRjOnpZrfGAr9Sewr3tIycdjYZD25lBDLRQi/N3vFSrdB5AuiwbHgsIjte+I5wudE0gaNHLdRSYX5Z47Ex9HHkC8CDEBuLYhJ8wzW3noPQsZAj+nFPEX3YR3hdc7ktCe9z2vfDz/ABww711dNylrPkmQWCHDFGjpJ4yTyreGFlm/Rccb29nQhCF2dEdb/kvib2pZra4TMe3j9c9Mzb/kviHalmtrhMz7eP1z1qIxYfCpb28HrWJmrA8l8R7EsticJlvbwOtYmPsi0IUOHmudQ1J0E/ZMhJ2jJMjMMN9c00N1OI1Gmo6QoqPkvDLmlr3NbVpeCS5z80hzQXE8RHGDpK79uIGscJ7kbcQNY4T3LKuc5Ny+cHb8GrSaOuOaIIAI5P8ATwj7jyrwlMloLGQ2uc97oYhgOrQ/0wA0Af2jTovvIrQrv24gaxwnuRtxA1jhPcg55zJ+FGfEdFLnB5Dg2tA0iHsed5zSvTxr3k7GgwmPY1u9iCjweOoIOilK5x0cqztxA1jhPcjbiBrHCe5ByOyZgE5xc8uDmv3zs4FzQwNJHm2NgupUVC7rGs5svBbBaa5tak8ZJqT9fstNuIGscJ7kbcQNY4T3IO9C4NuIGscJ7kbcQNY4T3IO9C4NuIGscJ7kbcQNY4T3IPHKLxG+v2FLXJ8LZ/Ib1oTF2xPQ4jWhhqQ6ugjiPKl0k+FM/kN60LUQy2Tvkj65+zV3zEEPbmu0XfQ1p9FDWPPw4bC15oc4nQTdQDi9C7tuIGscJ7lCzfZ6w7OhtNQOIgVvoCSd7XRe4rZ0k2jRVwzRmihpUXXHoC8NuIGscJ7kbcQNY4T3K8smeGPjTLbJhDW6fNTsW7pAZrWhxAa4u894cKDiHjLz24gaxwnuRtxA1jhPcnPL6nw8fo3hWZCaQQDcQR0LaPIMcSSXCt9xurm5tei5eW3EDWOE9yNuIGscJ7k5Zb3s+HjrWnpK2e1j84HizQOS8kn09gC61wbcQNY4T3I24gaxwnuUtt8rjjMZqO9C4NuIGscJ7kbcQNY4T3KNO9RmUXkh64+zl6bcQNY4T3LhtifhxGBrDU5wOgi6h5fSgXOPwp/8h/XFMpk74jvW7AlrmOFP/kP64pi7Hn4cNrg80JdXQTxAcS1UTceEHtLSSK8YND7io3c/A5DoIN+mpca6LvHdoppXttxA1jhPcjbiBrHCe5YuMvk1Hq+z2Fmx1cBUuuPG4kmtbiN8biFzwbDgMILQRm5tL9Gbm0v0/wBrVvtxA1jhPcjbiBrHCe5OMNRqbEgVzqGuc5+mvjFpIv4qtHoW0zY8F7y9wOcTXT/ta37NH15UbcQNY4T3I24gaxwnuTjPoajAsWBdvdBr62+Lhnct5PSuyWgNhtDG6Bo8wrWnoXJtxA1jhPcjbiBrHCe5JJDTvQuDbiBrHCe5G3EDWOE9yqu9C4NuIGscJ7kbcQNY4T3INbf8l8Q7Us1tcJmf5Efrnpj7WtCFEh5rTU1B0EfdLfbXCZn28frnrURxhxBBBIIvBFxBrcQeIqTGVM7zuJiXNZDA6Yl2kVBjQQQdBBiNBBHJRMLZ1hQ4rM7etvIpmA6FbR8DGVM7zuJiWd1E7zuJiTCbl4es35Y70bl4es35Y71NwL2MqZ3ncTEsbqZ3ncTEmF3Lw9Zvyx3o3Lw9Zvyx3psL2MqZ3ncTEg5UTvO4mJMJuXh6zfljvRuXh6zfljvTYXvdTO87iYkHKid53ExJhNy8PWb8sd6Ny8PWb8sd6bgXo5UzvO4mJAypnedxMQTC7l4es35Y70bl4es35Y702F7OVE7zuJiWRlRO87iYkwe5eHrN+WO9G5eHrN+WO9NhenZTTpBH7uLfdc6n1F6i4d2i6minFTkTHWnYcOG0Ehrg45pBYBcQfSl8lIbTMtYRvdma2nFm7IBToVlHTuonR/7uJiQMqZ3ncTEvvdm2FDisLt62hzaZgOgD0cq6ty8PWb8sd6mwvZyonedxMSN1M7zuJiTCbl4es35Y70bl4es35Y702F7OVM7zuJiRupnedxMSYTcvD1m/LHejcvD1m/LHemwvW6md53ExI3UzvO4mJMLuXh6zfljvRuXh6zfljvTYXrdTO87iYlndRO87iYkwm5eHrN+WO9G5eHrN+WO9NhejlTO87iYkbqZ3ncTEmF3Lw9Zvyx3o3Lw9Zvyx3psL5uonedxMSw7KedN37uJiTCbl4es35Y71y2lYUOEzO3rqkNpmAaQfTyJsLipMZTzooP3cS67xu0rnjwm/uHMoM0RnNAGjNEUinoomBsywocVpIDWhpzQAwG6g9Cto+DDKmd53ExIGVM7zuJiTC7l4es35Y70bl4es35Y71NhexlTO87iYkDKmd53ExJhNy8PWb8sd6Ny8PWb8sd6bC9DKmd53ExLIypnedxMSYTcvD1m/LHejcvD1m/LHemwvZyonedxMSN1M7zuJiTCbl4es35Y70bl4es35Y702F7OVE7zuJiWDlTO87iYkwu5eHrN+WO9G5eHrN+WO9NhehlRO87iYlk5UTvO4mJMJuXh6zfljvRuXh6zfljvTYXsZUTvO4uJRnKa1NbybySbySdJPnTHWhYMOEzO3rrwKZgGlL3azA2YjtaAA2NGaANAAiuAA9wCsozYfCZb28DrWpmrA8l8TuxLLYfCpb28DrmJmrA8l8R7FMhIoXnMR2Q2Oe9wa1gLnE6AAKkqneEiXrnft4+w52bsuaM2vo7K18yyq6oUDb2VcGWbBfmviiPUw9ioSRRprfy5wXnYeVgmYohCWmIdQTnRG0bdxV5UFiQqxMZaQmvd/RiGDDibC+OKZjYnJStSBy/8AClrZtYQNjAZskSM8Q4bAaVre5xN9GtFSTRBIoQhAIQhAIQhBFZReI31+wpapPhbP5DeuCZXKLxG+v2FLXJ8KZ7dvXBaiGWye8kfXP+LVJqMye8kfXP2apJzgASTQC8k6ABxlZVlCrb8sYV7hBiuhA5pihu9/79V32nb8GDDY+9+yXsDLy4XX+i8dKzyjtf6fqyycfKVQoeyMoYcd5hFj4UQCua8UJHm7l02za0OWZnvqamgaNJPHSvIrymts3o5zLhru70LwkJoRYbIgBAeA4A6RXlXuqxZZdUIQhECEIQCjMovJD1x/i5SajMovJD1x9nIFomOFP/kP64plcnfEd6/YEtcfhT/5D+uKZTJzxHev2BaqJVC8pyYENjohBIaC6g0mg0BV45Zwv/hi9A71zyzxx8pcpPKzIUNHyihtgQ45Y8iIc0NFM7j7vqsSGUQixGw9gitzq3ubQC4m8+5TnjvRyiaQsPeACSaAAkk8QGkqunLCFWuxRdjrTZKXdHZpVyzmPlblJ5WNCjbTtuDBY15JdsniBt5dovHmvHSvKysoIcZ5hFj4cSlc14pUebzpzx3racpvSXQoCdyqhse5jIb4mZ47mjeimm/tuUvZ86yNDESGatPSDxgjlSZ426izKV0IQhaVHW/5L4m9qWa2uEzP8iP1z0zNv+S+IdqWa2eEzP8AIj9c9axRiw+FS3t4HWtTNWB5L4j2JZbE4VLe3gdaxM1YHkviPYmQ9rXs5kzBfAeXBrwAS00dQEG4mvIqLbc5DiQBZNnsMcjNa5/9jA1wdUvFxNRedGmlTcrnlHJRo8tEhQXhj3gNzjWgaSM4XX3tqPeqfZmStqyrMyDNwIbSdQGpN17nMJPSor2tDJecESSbAcykrCzdkfeA81znCHpNKNIXrZlsT0vPskpuI2M2K3OY8NDSLnEXAC6rXCl/EaradyYm4hgTkKZZ+6YwBz6f04ovo4Zope11NFCKaF62Fk5GM3+7nY7IsdoIYyHoaBVpPFoziKU0k1vQckzknNlkaUDoX7aNH2cxCTsjW1DizNpStWi+qyLYbsv74w3RA537WShtoC4V38SpuAcRp5AFZspJSPGgmDBIbshDHuJoWwj45ZyuIu9647dyfc+HL/tixj5V7XQw6uYQ0ULHUvAIAv8AMg6rAtsTOyNdDdCiwXZkSG4h1CbwQ4XEG+/zKWUHk3Y8WC+PHjuYYsw8OcIdcxoaCGtBN50m9TigEIQgEIQgisovEb6/YUtcpwpn8hvXBMplF4jfW7Clrk+Fs/kN64LUQy2T3kj65/xatMrHkSkamqB7i4A/Qlb5PeSPrn/Fq67RZDdCe2KQGFpDiTSgN2nlWMvDp0rxzxvvFLlv3rZIRWmFsLR5MtDs5ude51dNTfSq2gzLY07JuDQ1phgho8UEGLWnxD6BexsWaEP9uJqFsBGdU+NmE1r6pPnouy1LFhhkEwY7IcSXBAc4ihoRXOPEQ4//AGI41y1X071Onuzc3d95L4s9fd45QPDLRln6N7V1OQF9fpVRFoz8KZbGjxIgD6BsCFxtbnCrjxZxFfr5lP2bZL3Rv3E1FhvdmgNa2mbR29B4rjUgcpK95+xpWNCiQ4DYAiXCraVac4ac28aCFbLWMOr08LjL6am5+u79nvknMsfLQ2tcCWNa1w5DTQVMKKseDBloLWl8Jp8VzgQA57bjedJ+qkXR2BpcXNDRUEkigpcaniXSeHh6s3nbj429ELURW5udnDN051bqctdCyxwIqDUHjF46VXLTKEIQCjMovJD1x9nKTUZlF5IeuPs5AtMfhT/5D+uKZTJ3xHev2BLVG4U/+Q/rimVyd8R3r9gWqiVVat6IY8xDk2ne3Pi05BeG/bEORWOJEa0EuIAGkk0A9JKhLJs4iYizRiMe1+cGlprQVGk6Lg0D3Ljn31P5pMu/ZplFZkaI6AIIaGwyTeaAGrc3e8YFCudtoTcvMw4UeI2KyLcCGhpBJpxecj3Fd1u2M+M5kaC/Miw9BOgitR9zy6Vzytix3Rmx5uI12x+K1tzRS+puHpWLLy7f8Zsu+zvyncRKxqatPcXAH6VUZAht2rNwpsbj8VSfutrIZEixZqHEiGLCNWggktGcTVrSeMAjQuN9iTTWGW/cwxBvea3PDK1JI1a+eiW2/m16aS9+6OstxdGkg7QGmlfM+JT7DoUtlGc2elnN8an0qe8r1tKy4LmQhBjsZElwKEuGi41dyX0NfP516WTZUV8YTUeKyIWijNjvaNI0i7jPSszG/L+iSXw8cgGgwYhN5MQg+cZrdPSVjIA/04o4g+7CP+Fs6wZqE6IJWK1sOIakO0tryXHl03KXsKyxLQhDBqa1cdFSezQtYY3c3PDWMvb2SCEIXd0R1v8AkviHalltof6mZ9vH656Zq3/JfEO1LLbPCZn+RH6561ijNicKlvbwOtYmZsDyXxO7Esth8KlvbwOtamYsKI0QryBvjpPoTIe9rybo0Iw2vzCS0kkVqAQSKedQrMmIlGh8YPLTCdnODiSYZh3Hf0pvXUuqM8+etj2Zms3pCNmbrDpCyqtHJeNmZn7onew2ioJoRDDYh0/3Oax/mIPLUbsyYiNrSML3FxBDqPLi0nZKOqbwdHm99i2Zms3pCNmZrN6QgibKsR8GIIjo7om9e12dWrqiDmk1NKt2OJxf+ofPWZWmzM1m9IRszNZvSEG6FpszNZvSEbMzWb0hBuhabMzWb0hGzM1m9IQboWmzM1m9IRszNZvSEEblF4jfW7Clrk+Fs9u3rkyWUDwWNoQd9xGvEUtsnwpnt29cFqIZbJ3yR9c/4tXRasoYsJzAQCaEVrSoNdLSCPSNC5Mn4jRDNSBvzpPmapPZmazekLNaxtl3EIzJ91avjZxLDDO9ABhlmbQcdc6jtPuvqsQsnS0OrEDnHMNSCKPrWI4FpBBcQDcbiONTmzM1m9IRszNZvSFnjHX8R1Pqi2WLvAHuDntdBc12boEJzXUA4id/f/v41mzbJfCc52yAktcBcbqvLwTUnRXQ2lVJ7M3WHSEbMzWb0hXjGfjZ6s2hYGTubRpiuLK5xuo4uMMw3XjiNx0V08q6hZObDzGuApFMVtRVvjVDXCt4odNeQqQ2Zms3pCNmZrN6QnGLetnfNQ0exIriP6rQALw1maDXPJFAdFXA/DfU3ros2ynwn52yVbm5tKHzUN5u0HRy8tSZHZmazekI2Zms3pCcYXrZ2aboWmzM1m9IRszNZvSFXJuozKLyQ9cfZykNmbrDpCjcoIjTDFCDvxoPmcgWqNwp/wDIf1xTK5O+I71+wJao3Cn/AMh/XFMnk+8BjqkDfcZ8wWskd1pSpiwnQwaF1L+ShB4qHiUe6xXO8aJxZt2cRSkQHxnEmue3Tq9EtszdYdIRszNZvSFzuMpZKiINiPDgXRSW72rRUAgU3unRcaek+/WWsWJQZ8Slx3oLqZ1HAOJrealvm3oUzszNZvSEbMzWb0hOETjEK6wolHARqAkGgBvNHVrfX+4H0tGji9rRsTZXh2yU3jYZuqXNBcXAnkNW4fOpTZmazekI2Zms3pCcIcYhHZOnNpsl9/FdQsa24VuO90+dSshKbEHCtc55dfpvpcTxnz+he2zM1m9IRszNZvSEmMiySN0LTZmazekI2Zms3pC0rdC02Zms3pCNmZrN6Qg4bf8AJfE3tSzW1wmZ9vH656Ze3YjTCuIO+Gg+lLRbXCZn28frnrWKOWXjFj2vad8xzXN9LSHD6gK1D9RJ7UlsD/yKolbBaFr8Is9qS2B/5FnwiT2pLYH/AJFUQsjtU0LaP1EntSWwP/IjwiT2pLYH/kVTWE0LYf1FntSWwP8AyLJ/UWe1JbA/8iqJWU0LZ4RZ7UlsD/yI8Is9qS2B/wCRVIafcgpoW3wiz2pLYH/kWR+ok9qS2B/5FUVs1NC2eESe1JbA/wDIgfqJPaktgf8AkVTKAmhbD+oc/qS4+B/bEVUhRXNcHg74ODq/7gc6tPSsrQKi2H9RZ7Ul8D/yIH6iz2pLYH/kVScsj/vSpoW3wiT2pLYH/kR4RJ7UlsD/AMiqXF7+1ZCaFs8Ik9qS2B/5EeESe1JbA/8AIqmVqO5NC3eESe1JbA/8iwP1EntSWwP/ACKprUJoW/wiT2pLYH/kWfCJPakt8t/5FUB/3pWxTQth/USe1JbA/wDIjwiT2pLYH/kVRd/3pWQmhbR+ok9qS2B/5EH9RJ/Ul/lv/Iqm1YTQ2MVxcXk74uLyf9xOcTT0lWrwiT2pLn0sfX6Poqisqi3eESe1JbA/8ix4RZ7UlsD/AMiqi1CmhbvCJPaktgf+RA/USe1JbA/8iqQ7VlNC2eESe1JbA/8AIseEWe1JbA/8iqYWCmhbfCLPaktgf+RHhFntSWwP/IqmsDT7k0Lb4RZ7UlsD/wAiPCLPaktgf+RVE6Uf8JoW8fqJPaktgf8AkR4RJ7UlsD/yKptQU0LYP1EntSWwP/IqtGiue5z3Xue5z3H/AHOcXOu9JK0C2Ko//9k=)

cryogenian commented:
Ok, I'm going to change formatter to `purescript-formatters` stuff. And put link to documentation under that field.

cryogenian commented:
![form](https://cloud.githubusercontent.com/assets/10245930/18389209/8b1c26a8-76ad-11e6-878b-e1251611592c.png)
![table](https://cloud.githubusercontent.com/assets/10245930/18389210/8b1ee7d0-76ad-11e6-84ea-2d210ac457ad.png)

beckyconning commented on src/SlamData/Workspace/Card/ChartOptions/Component.purs:
```diff
@@ -1,5 +1,5 @@
 {-
-Copyright 2016 SlamData, Inc.
+пCopyright 2016 SlamData, Inc.
```
Accidental п here.

beckyconning commented on src/SlamData/Workspace/Card/ChartOptions/Form/Component.purs:
```diff
@@ -319,7 +319,7 @@ render state = case state.chartType of
 
 eval ∷ Query ~> FormDSL
-eval (SetConfiguration conf next) = do
+eval (Load conf next) = do
```
Why the name change?

cryogenian replied:
For consistent with other `ChartOptions.Component` children

beckyconning commented on src/SlamData/Workspace/Card/ChartOptions/Component/CSS.purs:
```diff
@@ -77,3 +77,6 @@ funnelChartAlignParam = className "funnel-chart-align-param"
 gaugeChartIcon = className "gauge-chart-icon"
+
+metricChartIcon ∷ ClassName
+metricChartIcon = className "metrci-chart-icon"
```
typo here

beckyconning commented on src/Utils/DOM.js:
```diff
@@ -49,6 +49,33 @@ var getTextWidth_ = function(text) {
+    div.appendChild(span);
+    div.appendChild(block);
+
+    document.body.appendChild(div);
```
Is this necessary? If so I think this should be effectful.

cryogenian replied:
It's effectful in same sense as `getTextWidthPure` because `DOM` effect is erased with `div.remove()`

beckyconning replied:
Other things could be checking for the addition of elements to the body though which could make this work strangely. Does this not work if the element is not appended to the body?

cryogenian replied:
I don't think it works w/o appending, but let me check.

cryogenian replied:
Nope, doesn't work w/o appending to body :(

beckyconning replied:
Thats a shame. Given that its not possible without appending to the body I think even conceptually this is effectful. It seems like the places you are using this (i.e. not in render) can deal with effects so I'd like to see the type signature include an `Eff`. I'm happy to put this up to a second opinion though @natefaubion.

beckyconning commented on src/SlamData/Workspace/Card/ChartOptions/Metric/Component.purs:
```diff
@@ -0,0 +1,215 @@
+    , HH.input
+        $ [ HP.classes [ B.formControl ] ]
+        ⊕ foldMap (pure ∘ HP.value) state.formatter
+        ⊕ [ HP.placeholder "Will be replaced by actual aggregated value" ]
```
I feel like the placeholder attribute is for hints to users rather than for developers.

beckyconning replied:
Is essentially a loading state?

cryogenian replied:
Sorry? Maybe it's better to just remove this placeholder et al? There's a guide under this input

beckyconning replied:
Yeah I agree.

beckyconning commented on src/SlamData/Workspace/Card/Chart/MetricRenderer/Component.purs:
```diff
@@ -0,0 +1,220 @@
+  labelHeight ∷ Int
+  labelHeight =
+    Int.floor
+    $ fromMaybe zero
```
inconsistent indentation with below, either add double space on lines after `Int.floor` here (my preference) or remove the double spaces before lines after `C.top` below.

beckyconning commented:
Looks great other than the comments I've made : ). Thanks @cryogenian this will be really useful.

I kinda had a feeling that this might be cool as its own card type "build aggregation" with a "show aggregation" or something more generic like a "show single value" but I think taking advantage of the build chart card makes a lot of sense right now.

beckyconning commented:
Oh the other thought I had from using this is that it might be nice to find the largest possible value / label font sizes that fit across all values? So that all the font sizes in a multiple metric display would be consistent. Thats something we could possibly add later though.

cryogenian commented:
+ fixed typos
+ fixed indentation 

We are going to add card types for every chart, I think. So, there will be something like `Query -> Setup Chart -> Setup Metric -> Show Metric`

Do you mean the largest font size for all draftboard or for particular deck? I can add both (or some of this option) in this or separate pr, this is very straightforward stuff.

beckyconning commented:
I was thinking such that a deck like this one you posted above

![](https://cloud.githubusercontent.com/assets/10245930/18389210/8b1ee7d0-76ad-11e6-84ea-2d210ac457ad.png)

Is styled with the same font sizes across all metrics.

cryogenian commented:
@beckyconning Oh! Even not maximum font size, but all font sizes across all metrics, right? Then we need to utilize bus or manual setting of font sizes.

beckyconning commented:
Right I think this can be done in another PR but yes, I think it would look far cleaner if these all had the same font size, kinda like the smallest (largestFontSizeThatFitsMetricInCard <$> metrics <*> cards).

natefaubion commented:
It would have to be global coordinated since you can have arbitrarily nested draftboards. Honestly, I'd just pick fixed font sizes based on the responsive classes that get added to decks.

cryogenian commented:
I found classes depending on width, but not on height. `xxlarge` could be too small to handle even 24pt. Are you talking about adding height classes too?

beckyconning commented:
Yeah thats a good point if we use the responsive classes rather than working out the maximum size that will fit we won't need `getTextHeightPure`. 

Additionally I just remembered that "text height" is equivalent to the styled line-height (or font height) times the number of lines so it doesn't need to be calculated using the DOM like this especially if you set the height in the purescript source. It's only the width of text that changes due to proportional fonts.

cryogenian commented:
Unfortuantely setting `line-height` doesn't change font size. It just changes height of the line but actual text could be really small

natefaubion commented:
Well, its a trade off. Having responsive classes for width and height means you need configuration for every item in the product (6 x 6 = 36 configuration states). Most "heights" for this case are going to be relatively consistent, I'd wager. Having an extremely wide and short variation of this viz will likely look bad no matter what you do (_shrugs_). Dynamically sizing text is nice, but it leads to a very chaotic arrangement, and making them globally consistent requires non-trivial machinery.

beckyconning commented:
I mean the line-height is the actual measure of whether it will fit in the container's height but unless you set it yourself it is set based on the font-size. Font-size is the height of the text for a single line.

cryogenian commented:
Ok, how about this plan? 
+ Change `getTextHeightPure` to effectful `getTextHeight` 
+ Set maximum fontSize for every rendered value and label. (Like 120pt or something else)

cryogenian commented:
Oh! Maybe something like rounding fontSizes? like 16, 20, 24, 32, 48, 60, 72, 96?

beckyconning commented:
`ubuntuHeight state.valueFontSize state.value` == `state.valueFontSize` for single lines.

garyb commented:
I'm just swooping in with minimal context here, but seeing the above ^ we can't rely on some hard coded Ubuntu metrics, as the font can be changed in the (upcoming) themes or the custom stylesheet.

beckyconning commented:
@garyb thats a good point

beckyconning commented:
I'd prefer to see that fixed in another issue / pr though as we're assuming Ubuntu elsewhere with getTextWidth. We need to get the current font-face and font-style from the DOM before calculating any text measurements.

cryogenian commented:
`Slam` is actually `ReaderT` can we put it there? 
Oh! Once we switched to `purescript-css` from less we could just use config module!

natefaubion commented:
I don't know if we will switch to `purescript-css` because of themes. We have to assume that any of these things will be changed by a user-provided stylesheet.

beckyconning commented:
Not unless we assume the font face and style are consistent across the application. Even with this unless we've actually rendered the element before calculating it we can't be sure of the styles. Sorry this is complicated.

beckyconning commented:
My preference for now is to use a font-size based on the responsive classes with `text-overflow: ellipsis;` to show the author that they need to increase the size of the deck or consumer that they need to rotate their device or increase their window size.

In the future we could use something like https://github.com/adactio/FitText.js which is a non-jquery alternative to fittext.js to adjust the sizes after rendering them.

beckyconning commented:
Basically I think to do this accurately we need to adjust text sizes after they are rendered. So I'd like to see us do that or not try.

beckyconning commented:
Sorry this is getting complicated @cryogenian its a really good pr regardless of this confusing issue.

cryogenian commented:
Actually, we can't just set font based on responsive layout. Because there could be different length of value and label strings. And if we use ellipsis to notify user about size of metric that 
1) seems a bit unfriendly, because user has set size of that deck already
2) could change responsive class. 

That's why I think it'll be a bit better to have distinct set of font sizes and round to them. 

@beckyconning  nope, `ubuntuHeight state.valueFontSize state.value /=  state.valueFontSize` :( 

How can we calculate text size after text was rendered? Just render it once and then approximate width and height as linear, right? It will blink screen with text at least once.

natefaubion commented:
I don't know if it will necessarily blink. It might reflow, but not repaint.

cryogenian commented:
+ removed placeholder
+ used `FitText`-like approach with discreate set of available sizes.

cryogenian commented:
@becky Is this good to go now?

beckyconning commented:
Reading now, thanks : ).

beckyconning commented:
(Accidentally closed)

beckyconning commented on src/Utils/DOM.purs:
```diff
@@ -59,8 +59,14 @@ foreign import centerPopupWindowFeatures ∷ ∀ eff. Int → Int → Window →
-  ∷ {fontSize ∷ Int, fontFamily ∷ String, fontStyle ∷ String, text ∷ String} → Number
+-- | Takes an array of (div) elements, array of available font sizes, maximum dimension and
+-- | returns sizes of fitted elements
+foreign import fitTexts
```
Why do you need the size of the elements?

cryogenian replied:
To calculate `top` from size of container and this texts

beckyconning replied:
Is that to vertically centre the text?

cryogenian replied:
Yes

beckyconning commented on src/SlamData/Workspace/Card/Chart/MetricRenderer/Component.purs:
```diff
@@ -0,0 +1,128 @@
+  pure $ continue $ if (state.labelHeight + state.valueHeight) > state.height then Low else High
+eval (SetValueElement mbEl next) = do
+  for_ mbEl \el → H.modify _{valueElement = Just el}
+  adjustFontSizes
```
I think its not guaranteed that the element is rendered at this point. This may have changed since I last checked though, @garyb?

garyb replied:
The element exists at this point, but it has not been added to the DOM yet.

beckyconning commented on src/SlamData/Workspace/Card/Chart/MetricRenderer/Component.purs:
```diff
@@ -0,0 +1,95 @@
+eval (SetElement mbEl next) = do
+  for_ mbEl \el → do
+    H.modify _{element = Just el}
+  adjustFontSizes
```
I think it's luck that this works here, see https://jsfiddle.net/2o1ckmse/.

Make this a lifecycle component and use initialiser. It might also be worth using adding a key (or some keys) to prevent the element(s) being replaced.

cryogenian replied:
Actually this `adjustFontSizes` didn't work :) I've removed it. We know that `SetDimensions` and `SetMetric` fires after element is available, so, won't this be enough?

beckyconning replied:
Yep! 👍

beckyconning commented:
Looks good : )
